config/management/operational/src/validator_state.rs

```diff
@@ -2,67 +2,119 @@
+    pub in_validator_set: Option<bool>,
+
+    /// Check if the validator network key held in secure storage matches
+    /// that registered on-chain.
+    pub validator_network_key_match: Option<bool>,
 }
 
 impl VerifyValidatorStateResult {
-    pub fn is_state_consistent(&self) -> bool {
-        self.in_validator_set == Some(true) && self.consensus_key_match == Some(true)
+    pub fn is_valid_state(&self) -> bool {
+        self.in_validator_set == Some(true)
+            && self.consensus_key_match == Some(true)
+            && self.consensus_key_unique == Some(true)
+            && self.validator_network_key_match == Some(true)
+            && self.fullnode_network_key_match == Some(true)
     }
 }
 
 #[derive(Debug, StructOpt)]
 pub struct VerifyValidatorState {
+    #[structopt(flatten)]
+    config: ConfigPath,
     #[structopt(long, required_unless = "config")]
     json_server: Option<String>,
     #[structopt(flatten)]
-    validator_config: diem_management::validator_config::ValidatorConfig,
+    validator_backend: ValidatorBackend,
 }
 
 impl VerifyValidatorState {
     pub fn execute(self) -> Result<VerifyValidatorStateResult, Error> {
         // Load the config, storage backend and create a json rpc client.
         let config = self
-            .validator_config
-            .config()?
-            .override_json_server(&self.json_server);
+            .config
+            .load()?
+            .override_json_server(&self.json_server)
+            .override_validator_backend(&self.validator_backend.validator_backend)?;
         let storage = config.validator_backend();
         let client = JsonRpcClientWrapper::new(config.json_server);
         let owner_account = storage.account_address(OWNER_ACCOUNT)?;
 
-        // Verify if the validator is in the set
-        let in_validator_set = client
-            .validator_set(None)?
-            .iter()
-            .any(|vi| vi.account_address() == &owner_account);
+        let validator_infos = client.validator_set(None)?;
 
-        // TODO(khiemngo): consider return early if the validator is not in the set
+        let mut result = VerifyValidatorStateResult {
+            consensus_key_match: None,
+            consensus_key_unique: None,
+            fullnode_network_key_match: None,
+            in_validator_set: None,
+            validator_network_key_match: None,
+        };
 
-        // Fetch the current on-chain config for this operator's owner
+        // Verify if the validator is in the set.
+        result.in_validator_set = Some(
+            validator_infos
+                .iter()
+                .any(|vi| vi.account_address() == &owner_account),
+        );
+
+        // Fetch the current on-chain config for this operator's owner.
+        // Check if the consensus key held in secure storage matches
+        // that registered on-chain.
         let validator_config = client.validator_config(owner_account).and_then(|vc| {
```

Comment:
One thing I'm not sure/want to ask is: even if this node is not in the validator set, but its account is registered/available on-chain, the call `client.validator_config(owner_account)` should be able to return the config (unless there are some unexpected errors (e.g., network))? @JoshLind

Reply:
@khiemngo, yeah, exactly. It should still work 😄